### PR TITLE
bpo-46968: Query altstack size dynamically on Linux kernel >=5.14

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
In Linux kernel 5.14 one can dynamically request size of altstacksize
needed to handle signals based on hardware capabilities. 
This patch sets stack size to be the maximum of current value of SIGSTKSZ * 2 and
SIGSTKSZ + getauxval(AT_MINSIGSTKSZ).

This changes allows for Python extension's request to Linux kernel
to use AMX_TILE instruction set on Sapphire Rapids Xeon processor
to succeed, unblocking use of the ISA in frameworks.

This change proposes a solution to the problem described 
in https://bugs.python.org/msg414809

https://bugs.python.org/issue46968

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@oleksandr-pavlyk

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by oleksandr-pavlyk** [comment]
The CLA has been signed. I am the first time contributor to CPython proper. Please guide through the process of getting this PR to the required state to make the merge possible. Thank you.

**Comment by AlexWaygood** [comment]
> The CLA has been signed. I am the first time contributor to CPython proper. Please guide through the process of getting this PR to the required state to make the merge possible. Thank you.

Thanks for signing the CLA! It'll take a few days for the bot to update, as the form needs to be manually reviewed by a human at the PSF :)

The first step is to change the title of your PR to "bpo-46968: Query altstack size dynamically on Linux kernel >=5.14". This will cause the bots to automatically link this PR to that BPO issue, which will make the relevant CI check go green.

The second thing you'll need to do is to add a NEWS entry to your PR, which you can do using https://blurb-it.herokuapp.com.

**Comment by AlexWaygood** [comment]
You'll probably also need to add a test to your PR to make it mergeable.

For a fuller guide to the process, the devguide is [here](https://devguide.python.org) :)

**Comment by oleksandr-pavlyk** [comment]
> You'll probably also need to add a test to your PR to make it mergeable.

Since size of size of sigaltstack used by faulthandler extensions is not user accessible to the best of my knowledge, I am not entirely sure how once can test the change. 

Suggestions are welcome.

**Comment by AlexWaygood** [comment]
> Suggestions are welcome.

I'm afraid this is an area I know next to nothing about, so I can't really help you any further :/

File: aclocal.m4 [hunk 1]
```diff
@@ -1,6 +1,6 @@
-# generated automatically by aclocal 1.16.3 -*- Autoconf -*-
```
Comment (gpshead) [inline aclocal.m4]: We need to use the same versions when regenerating these. I believe the version mismatch is why the generated configure script deletes some unrelated runstatedir lines as well.

https://devguide.python.org/setup/#regenerate-configure

**Review by gpshead — CHANGES_REQUESTED** [review]
The core of the changes look good, the annoying part is in the generated configure and aclocal.m4 files - those should be generated with the same version as the existing ones.

See the 2022 comment on https://bugs.python.org/issue44035 for a pointer to a docker image that can be used for the autoconf regeneration stuff.  (that image is not yet documented AFAICT)

File: configure.ac [hunk 2]
```diff
@@ -2405,6 +2405,8 @@ AC_CHECK_HEADERS([net/if.h], [], [],
 #endif
 ])
 
+AC_CHECK_HEADERS([linux/auxvec.h], [], [], [])
```
Comment (tiran) [inline configure.ac]: Move this header check to the large ``AC_CHECK_HEADERS`` around line 2376.

**Comment by oleksandr-pavlyk** [comment]
I have made the requested changes; please review again

I modified `configure.ac` as requested by @tiran 

I regenerated `configure` and `aclocal.m4` by rerunning `autoreconf` using

```
bash-4.4$ autoreconf --version
autoreconf (GNU Autoconf) 2.69
Copyright (C) 2012 Free Software Foundation, Inc.
License GPLv3+/Autoconf: GNU GPL version 3 or later
<http://gnu.org/licenses/gpl.html>, <http://gnu.org/licenses/exceptions.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David J. MacKenzie and Akim Demaille.
```

Assuming this change eventually gets accepted, what is the process of requesting backports to all earlier versions of CPython?

**Review by tiran — CHANGES_REQUESTED** [review]
You need a verify specific combination of autoconf 2.69 + runstatedir patch, autoarchive version, and pkg-config m4 macro version. Your version of autoconf aclocal macros is too old. If you cannot use the container image to regenerate the files, then apply the changes manually.

* revert ``aclocal.m4`` and ``configure`` changes.
* manually add ``linux/auxvec.h`` to ``configure``

**Comment by tiran** [comment]
You can find the container image and instructions at https://quay.io/repository/tiran/cpython_autoconf

**Comment by oleksandr-pavlyk** [comment]
I have made the requested changes; please review again

@tiran I used the cpython_autconf:269 docker to regenerate configure script. Thank you for patiently guiding me through the process.

File: Modules/faulthandler.c [hunk 3]
```diff
@@ -1372,7 +1379,18 @@ _PyFaulthandler_Init(int enable)
...
+        unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
+        if (at_minstack_size > SIGSTKSZ) {
+            stack.ss_size = SIGSTKSZ + at_minstack_size;
+        } else {
+            stack.ss_size = SIGSTKSZ * 2;
+        }
+    }
+#else
     stack.ss_size = SIGSTKSZ * 2;
+#endif
```
Comment (vstinner) [inline Modules/faulthandler.c]: Please add a comment to explain why you use AT_MINSIGSTKSZ. I suggest to change the code like that:

```suggestion
    stack.ss_size = SIGSTKSZ * 2;

#ifdef AT_MINSIGSTKSZ
    // bpo-46968: (...)
    unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
    if (at_minstack_size != 0) {
        stack.ss_size = SIGSTKSZ + at_minstack_size;
    }
#endif
```

File: Misc/NEWS.d/next/Core and Builtins/2022-03-09-20-10-34.bpo-46968.Rz-uPr.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Dynamically determine size of sigaltstack CPython allocates to ensure Linux signal delivery on the hardware running the interpreter. The required Linux kernel support is available since kernel version 5.14.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2022-03-09-20-10-34.bpo-46968.Rz-uPr.rst]: Please mention faulthandler somewhere.

File: Modules/faulthandler.c [hunk 5]
```diff
@@ -20,6 +20,19 @@
...
+#if defined(HAVE_SIGALTSTACK) && defined(HAVE_SIGACTION)
+#  define FAULTHANDLER_USE_ALT_STACK
+#endif
+
+#ifdef FAULTHANDLER_USE_ALT_STACK
+#ifdef HAVE_LINUX_AUXVEC_H
+#  include <linux/auxvec.h>
+#endif
+#include <sys/auxv.h>
+#endif
```
Comment (vstinner) [inline Modules/faulthandler.c]: It's unclear to me why <sys/auxv.h> is always included, but <linux/auxvec.h> is included conditionally. Would it be possible that AT_MINSIGSTKSZ is defined by <sys/auxv.h>?

```suggestion
#ifdef FAULTHANDLER_USE_ALT_STACK
#  ifdef HAVE_LINUX_AUXVEC_H
#    include <linux/auxvec.h>
#  endif
#  include <sys/auxv.h>
#endif
```

**Review by vstinner — COMMENTED** [review]
I like AT_MINSIGSTKSZ feature! It's better than coarse ``SIGSTKSZ * 2``. Here is a first review.

**Comment by tiran** [comment]
> @tiran I used the cpython_autconf:269 docker to regenerate configure script. Thank you for patiently guiding me through the process.

Sorry for the extra work! I created https://github.com/python/cpython/pull/31792 to reduce the pain for future contributors.

File: Modules/faulthandler.c (hunk 5 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: `linux/auxvec.h` does not exist on OSX, but without it `getauxval` (include from `auxv.h`) is not used, so I will move the `#include <sys/auxv.h>` inside the `HAVE_LINUX_AUXVEC_H` conditional.

File: Misc/NEWS.d/next/Core and Builtins/2022-03-09-20-10-34.bpo-46968.Rz-uPr.rst (hunk 4 above)
Comment (oleksandr-pavlyk) [reply Misc/NEWS.d/next/Core and Builtins/2022-03-09-20-10-34.bpo-46968.Rz-uPr.rst]: Done.

File: Modules/faulthandler.c (hunk 5 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: Changes applied.

File: Modules/faulthandler.c (hunk 3 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: Done. Thank you for the review.

**Comment by oleksandr-pavlyk** [comment]
I have cleaned up git history down to 3 commits: changes to `configure`, changes to `Modules/faulthandler.c` and a single entry to NEWS.

**Comment by oleksandr-pavlyk** [comment]
I have made the requested changes; please review again

File: Misc/NEWS.d/next/Library/2022-03-10-14-51-11.bpo-46968.ym2QxL.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+Dynamically determine size of sigaltstack CPython allocates during initialization of the faulthandler extension to ensure Linux signal delivery on the hardware running the interpreter. The required Linux kernel support is available since kernel version 5.14.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2022-03-10-14-51-11.bpo-46968.ym2QxL.rst]: I would prefer to put faulthandler at the beginning. Like:

```
:mod:`faulthandler`: On Linux 5.14 and newer, dynamically determine size (...) using ``getauxval(AT_MINSIGSTKSZ)`` (...).
```

Rather than "sigaltstack", I suggest to say: "signal handler stack size".

IMO it's worth it to go into the details:

"This changes allows for Python extension's request to Linux kernel
to use AMX_TILE instruction set on Sapphire Rapids Xeon processor
to succeed, unblocking use of the ISA in frameworks."

nitpick: Can you please try to limit to 80 columns?

File: Modules/faulthandler.c [hunk 7]
```diff
@@ -1373,6 +1378,17 @@ _PyFaulthandler_Init(int enable)
        signal handler uses more than SIGSTKSZ bytes of stack memory on some
        platforms. */
     stack.ss_size = SIGSTKSZ * 2;
+#ifdef AT_MINSIGSTKSZ
+    {
```
Comment (vstinner) [inline Modules/faulthandler.c]: These braces are not needed. Python 3.11 uses C11, see: https://peps.python.org/pep-0007/#c-dialect

File: Modules/faulthandler.c [hunk 8]
```diff
@@ -1373,6 +1378,17 @@ _PyFaulthandler_Init(int enable)
...
+        /* bpo-46968: Query Linux for minimal stack size to ensure signal delivery
+           for the hardware running CPython. This OS feature is available in
+           Linux kernel version >= 5.14 */
+        unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
+        if (at_minstack_size != 0) {
+            stack.ss_size = SIGSTKSZ + at_minstack_size;
+	}
```
Comment (vstinner) [inline Modules/faulthandler.c]: strange indentation :-)

File: Modules/faulthandler.c (hunk 7 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: I had a question here. As is the change will only be effective on the new hardware if CPython was compiled against kernel 5.14 of later. Would be acceptable to do

```c
#ifndef AT_MINSIGSTKSZ
#define AT_MINSIGSTKSZ 51
#endif
```

relying on the fact that `getauxval` returns 0 for unknown tokens?

This would permit backport of this fix to earlier version of CPython, right?

File: Modules/faulthandler.c (hunk 8 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: Addressed.

File: Modules/faulthandler.c (hunk 7 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: > These braces are not needed. Python 3.11 uses C11, see: https://peps.python.org/pep-0007/#c-dialect

Removed, and indentation adjusted. Would appreciate feedback about tweaking the proposed fix so that the call to `getauxval` is not in the preprocessor branch by defining `AT_MINSIGSTKSZ` if undefined.

File: Misc/NEWS.d/next/Library/2022-03-10-14-51-11.bpo-46968.ym2QxL.rst (hunk 6 above)
Comment (oleksandr-pavlyk) [reply Misc/NEWS.d/next/Library/2022-03-10-14-51-11.bpo-46968.ym2QxL.rst]: I modified the .rst file directly as requested.

**Comment by oleksandr-pavlyk** [comment]
I have made the requested changes; please review again

File: Modules/faulthandler.c (hunk 7 above)
Comment (vstinner) [reply Modules/faulthandler.c]: Please don't hardcode constants in Python: don't add ``#define AT_MINSIGSTKSZ 51``.

Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c]: > Please don't hardcode constants in Python: don't add `#define AT_MINSIGSTKSZ 51`.

Ok, understood, but had to ask.

File: Modules/faulthandler.c [hunk 9]
```diff
@@ -1373,6 +1378,15 @@ _PyFaulthandler_Init(int enable)
        signal handler uses more than SIGSTKSZ bytes of stack memory on some
        platforms. */
     stack.ss_size = SIGSTKSZ * 2;
+#ifdef AT_MINSIGSTKSZ
+    /* bpo-46968: Query Linux for minimal stack size to ensure signal delivery
+       for the hardware running CPython. This OS feature is available in
+       Linux kernel version >= 5.14 */
+    unsigned long at_minstack_size = getauxval(AT_MINSIGSTKSZ);
+    if (at_minstack_size != 0) {
+        stack.ss_size = SIGSTKSZ + at_minstack_size;
```
Comment (vstinner) [inline Modules/faulthandler.c:1387]: I'm surprised that AT_MINSIGSTKSZ doesn't return directly the value which should be passed to sigaltstack(), but it seems like your code is the correct way to use the value.

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
I wrote PR #31824 to add ``os.sysconf_names['SC_MINSIGSTKSZ']``.

File: Modules/faulthandler.c (hunk 9 above)
Comment (oleksandr-pavlyk) [reply Modules/faulthandler.c:1387]: The `AT_MINSIGSTKSZ` is a preprocessor variable. The `getauxval` checks hardware capabilities (e.g. total size of CPU state to be saved at context switches) and may return different values on different platforms.

Comment (vstinner) [reply Modules/faulthandler.c:1387]: I mean that I expected to write (simplified code without error handling): ``stack.ss_size = getauxval(AT_MINSIGSTKSZ)``, rather than ``stack.ss_size = SIGSTKSZ + getauxval(AT_MINSIGSTKSZ)``. But your code looks correct according to the doc that I found on the Internet. Like the official Linux documentation:

```
ss.ss_size = getauxval(AT_MINSIGSTKSZ) + SIGSTKSZ;
```

ref: https://docs.kernel.org/x86/elf_auxvec.html#introduction

**Comment by vstinner** [comment]
Merged, thank you!

**Comment by vstinner** [comment]
I fixed the conflict on configure.ac and backported the change to 3.10: https://github.com/python/cpython/pull/31830 Then I will backport the fix to 3.9.
